x-pack/plugins/endpoint/public/embeddables/resolver/view/submenu.tsx

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import React, { ReactNode, useState, useMemo, useCallback } from 'react';
+import { EuiSelectable, EuiButton } from '@elastic/eui';
+import styled from 'styled-components';
+
+/**
+ * i18n-translated titles for submenus and identifiers for display of states:
+ *   initialMenuStatus: submenu before it has been opened / requested data
+ *   menuError: if the submenu requested data, but received an error
+ */
+export const subMenuAssets = {
+  initialMenuStatus: i18n.translate('xpack.endpoint.resolver.relatedNotRetrieved', {
+    defaultMessage: 'Related Events have not yet been retrieved.',
+  }),
+  menuError: i18n.translate('xpack.endpoint.resolver.relatedRetrievalError', {
+    defaultMessage: 'There was an error retrieving related events.',
+  }),
+  relatedAlerts: {
+    title: i18n.translate('xpack.endpoint.resolver.relatedAlerts', {
+      defaultMessage: 'Related Alerts',
+    }),
+  },
+  relatedEvents: {
+    title: i18n.translate('xpack.endpoint.resolver.relatedEvents', {
+      defaultMessage: 'Events',
+    }),
+  },
+};
+
+interface ResolverSubmenuOption {
+  optionTitle: string;
+  action: () => unknown;
+  prefix?: number | JSX.Element;
+}
+
+export type ResolverSubmenuOptionList =
+  | ResolverSubmenuOption[]
+  | 'waitingForRelatedEventData'
+  | typeof subMenuAssets.initialMenuStatus;
+
+const OptionList = React.memo(
+  ({
+    subMenuOptions,
+    isLoading,
+  }: {
+    subMenuOptions: ResolverSubmenuOptionList;
+    isLoading: boolean;
+  }) => {
+    const selectableOptions =

```

Comment:
Since `selectableOptions` is only used on the first render, you can calculate it in a function that you pass to `useState`:
```
const [options, setOptions] = useState(() => typeof subMenuOptions !== 'object'
    ? []
    : subMenuOptions.map((opt: ResolverSubmenuOption): {
        label: string;
        prepend?: ReactNode;
    } => {
        return opt.prefix
            ? {
                label: opt.optionTitle,
                prepend: <span>{ opt.prefix } < /span>,
            }
            : {
                label: opt.optionTitle,
                prepend: <span />,
            };
    })
);
```
This way, you won't allocate that array of jsx on each render (which is good because its only used once) and it should be more clear to the reader that `selectableOptions` is only used to initialize the state on the first ender.

Reply:
Good catch - thanks. https://github.com/elastic/kibana/pull/64121/commits/91f2bd61331f950098762915dbf9c408b9ab0128